docs/source/models.rst

```diff
@@ -922,7 +922,90 @@ and behavior:
    :width: 69%
 
-More information about model evaluation behaviors and outputs is available in the
+
+Evaluating with Custom Metrics
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If the default set of metrics is insufficient, you can specify a list of ``custom_metrics`` functions to 
+:py:func:`mlflow.evaluate()` to produce custom performance metrics for the model(s) that you're evaluating. Custom metric
+functions should accept two arguments: a DataFrame containing ``prediction`` and ``target`` columns, 
+and a dictionary containing the default set of metrics. For a full list of default metrics, refer to the documentation 
+of :py:func:`mlflow.evaluate()`.
+
+The following `short example from the MLflow GitHub Repository
+<https://github.com/mlflow/mlflow/blob/master/examples/evaluation/evaluate_with_custom_metrics.py>`_ 
+uses :py:func:`mlflow.evaluate()` with a custom metric function to evaluate the performance of a regressor on the
+`California Housing Dataset <https://www.dcc.fc.up.pt/~ltorgo/Regression/cal_housing.html>`_.
+Note that custom metric functions can return both metrics and artifacts. They can either return a single
+dictionary of metrics, or two dictionaries representing metrics and artifacts.
+
+.. code-block:: py
+
+    from sklearn.linear_model import LinearRegression
+    from sklearn.datasets import fetch_california_housing
+    from sklearn.model_selection import train_test_split
+    import numpy as np
+    import mlflow
+    import tempfile
+    import os
+    import matplotlib.pyplot as plt
+
+    # loading the California housing dataset
+    cali_housing = fetch_california_housing(as_frame=True)
+
+    # split the dataset into train and test partitions
+    X_train, X_test, y_train, y_test = train_test_split(
+        cali_housing.data, cali_housing.target, test_size=0.2, random_state=123
+    )
+
+    # train the model
+    lin_reg = LinearRegression().fit(X_train, y_train)
+
+    # creating the evaluation dataframe
+    eval_data = X_test.copy()
+    eval_data["target"] = y_test
+
+
+    def example_custom_metric_fn(eval_df, builtin_metrics):
+        """
+        This example custom metric function creates a metric based on the ``prediction`` and
+        ``target`` columns in ``eval_df`` and a metric derived from existing metrics in
+        ``builtin_metrics``. It also generates and saves a scatter plot that visualizes the
+        relationship between the predictions and targets for the given model to a file as an
+        image artifact.
+        """
+        metrics = {
+            "squared_diff_plus_one": np.sum(np.abs(eval_df["prediction"] - eval_df["target"] + 1) ** 2),
+            "sum_on_label_divided_by_two": builtin_metrics["sum_on_label"] / 2,
+        }
+        plt.scatter(eval_df["prediction"], eval_df["target"])
+        plt.xlabel("Targets")
+        plt.ylabel("Predictions")
+        plt.title("Targets vs. Predictions")
+        plot_path = os.path.join(tmp_dir, "example_scatter_plot.png")
```

Comment:
Can we create tmp_dir inside the custom metric function in this example? In general the function should not access anything other than eval_df and builtin_metrics, to make pickling the function much easier. This would be useful for distributed processing in the future.

Reply:
Uhmm, yeah, I did that originally. However, the problem with having the tmp_dir context inside the function is that once the function finishes execution, it deletes the directory. This is problematic because later on in our code, we will need to reference to those produced artifacts. Perhaps we could introduce another argument to custom metric functions called `tmp_dir`.

Reply:
I see. In this case, I think we need to change the API to provide this directory as part of the custom metric function argument, since we do expect users to write data to disk, in addition to the tuple of dicts returned.